add data availability docs (#1995)
OffchainLabs/arbitrum-docs: arbitrum-docs/how-arbitrum-works/12-data-availability.mdx

---
title: Data Availability
description: 'Learn how data availability works in arbitrum'
author: JasonWan
sme: JasonWan
user_story: As a current or prospective Arbitrum user, I need to learn more about how data availability works on Arbitrum.
content_type: get-started
---

# How Arbitrum data availability works

## What is the general view of Arbitrum data flow?

Arbitrum currently supports two primary data availability mechanisms:

**Rollup Mode:** In this mode, all transaction data is included in either the calldata of transactions submitted to the parent chain (e.g., Ethereum mainnet for Arbitrum One) or the blobs submitted by the transaction. This inclusion ensures that all data is readily available on-chain for anyone to download and verify.

> **Anytrust Mode:** In Anytrust mode, transaction data initially gets submitted to a group of nodes known as the Data Availability Committee (DAC). The DAS stores and distributes the data. Instead of including the entire dataset on-chain, only only a cryptographic proof that the data has been stored by the DAC (Data Availability Certificate, or DACert) is submitted to the parent chain. This proof significantly reduces the amount of data stored on-chain, reducing costs.

Reviewer: ```suggestion
**Anytrust Mode:** In Anytrust mode, transaction data initially gets submitted to a group of nodes known as the Data Availability Committee (DAC). The DAC stores and distributes the data. Instead of including the entire dataset on-chain, only a cryptographic proof that the data has been stored by the DAC (called a Data Availability Certificate, or DACert) is submitted to the parent chain. This proof significantly reduces the amount of data stored on-chain, reducing costs.
```

Because of those data availability mechanisms, Arbitrum Nitro nodes synchronize their data differently than Ethereum nodes or other layer-one network nodes. While Go-Ethereum nodes utilize a sophisticated P2P network to synchronize with the Ethereum blockchain by discovering other nodes, exchanging data, and participating in the consensus mechanism, Arbitrum nodes diverge from this traditional approach and use a trustless process.

Here's how Arbitrum data flow works:

1. Batching and submission:
   1. The sequencer queues transactions and batches them together.
   2. These batches get submitted to the parent chain:

>       1. In Rollup mode, the Sequencer submits the batch of transactions directly to the sequencer inbox contract on the parent chain. (Blob or calldata directly)

Reviewer: Some of the occurrences of "sequencer" are still in uppercase, can you update those too ?🙏

>       2. In Anytrust mode, the Sequencer sends the batch to the the Data Availability Committee (DAC) and then submits the Data Availability Certificate (DACert) which is returned and generated by the DAS to the parent chain.

Reviewer: ```suggestion
      2. In Anytrust mode, the Sequencer sends the batch to the Data Availability Committee (DAC) and then submits the Data Availability Certificate (DACert) which is returned and generated by the DAS to the parent chain.
```

> 2. Node Synchronization:

Reviewer: ```suggestion
2. Node synchronization:
```
Can you update the formatting of the other points in this section to be "Sentence-case"? 🙏

**Anytrust Mode:** In Anytrust mode, transaction data initially gets submitted to a group of nodes known as the Data Availability Committee (DAC). The DAS stores and distributes the data. Instead of including the entire dataset on-chain, only a cryptographic proof (Data Availability Certificate, or DACert) is submitted to the parent chain. This proof significantly reduces the amount of data stored on-chain, reducing costs.

Because of those data availability mechanisms, Arbitrum Nitro nodes synchronize their data differently than Ethereum nodes or other layer-one network nodes. While Go-Ethereum nodes utilize a sophisticated P2P network to synchronize with the Ethereum blockchain by discovering other nodes, exchanging data, and participating in the consensus mechanism, Arbitrum nodes diverge from this traditional approach.

However, Arbitrum nodes do **not** primarily rely on a traditional peer-to-peer (P2P) mechanism for syncing their state as many other blockchains do.

Here's how Arbitrum data flow works:

> 1. Batching and submission:

Reviewer 2: For these lists below, I would change to:
1. asdf
  a. asdf
    i.

---
title: Data Availability
description: 'Learn how data availability works in arbitrum'
author: JasonWan
sme: JasonWan
user_story: As a current or prospective Arbitrum user, I need to learn more about how data availability works on Arbitrum.
content_type: get-started
---

# How Arbitrum data availability works

## What is the general view of Arbitrum data flow?

Arbitrum currently supports two primary data availability mechanisms:

> **Rollup Mode:** In this mode, all transaction data is included in either the calldata of transactions submitted to the Ethereum mainnet (the parent chain) or the blobs submitted by the transaction. This inclusion ensures that all data is readily available on-chain for anyone to download and verify.

Reviewer: ```suggestion
**Rollup Mode:** In this mode, all transaction data is included in either the calldata of transactions submitted to the parent chain (e.g., Ethereum mainnet for Arbitrum One) or the blobs submitted by the transaction. This inclusion ensures that all data is readily available on-chain for anyone to download and verify.
```
This document should also consider Orbit chains, so we can change some bits to generalize the parent chain used.

> **Anytrust Mode:** In Anytrust mode, transaction data initially gets submitted to a group of nodes known as the Data Availability Servers (DAS). The DAS stores and distributes the data. Instead of including the entire dataset on-chain, only a cryptographic proof (Data Availability Certificate, or DACert) is submitted to the parent chain. This proof significantly reduces the amount of data stored on-chain, reducing costs.

Reviewer: Does it make sense to say Data Availability Committee (DAC) instead? It involves the committee as a whole, instead of the individual servers.

Reviewer: Also, I'd say "only a cryptographic proof that the data has been stored by the DAC (Data Availability Certificate...)", so it is not confused with other types of proof.

Author: Sorry, kind confused here, do you here mean DACert or DAC?

Reviewer: Oh, my comment is a bit confusing, you're right. I meant that the data is stored by the DAC, and the cryptographic proof is the Data Availability Certificate. So the text would read "only a cryptographic proof that the data has been stored by the DAC (Data Availability Certificate, or DACert)...". I think we've introduced the DAC before, so in this case we are just saying that the data is stored by the DAC, and they return the DACert, which is posted to the parent chain.
Let me know if it's clearer now. If not, maybe we can try to find a different way to describe the process 🙏

Because of those data availability mechanisms, Arbitrum Nitro nodes synchronize their data differently than Ethereum nodes or other layer-one network nodes. While Go-Ethereum nodes utilize a sophisticated P2P network to synchronize with the Ethereum blockchain by discovering other nodes, exchanging data, and participating in the consensus mechanism, Arbitrum nodes diverge from this traditional approach.

> However, Arbitrum nodes do **not** primarily rely on a traditional peer-to-peer (P2P) mechanism for syncing their state as many other blockchains do.

Reviewer: I'd remove this paragraph. It doesn't add value to what has already been said, and it says that it somehow relies on a P2P mechanism (not primarily, but still...)

Author: I add this paragraph is because I want to say arbitrum sync and process the data by a much more trustless way, but ya, I forget to add this word in, how do you think if we add trustless in this paragraph?

Reviewer: Maybe in the paragraph above that compares Ethereum to Arbitrum: "However, Arbitrum nodes do not primarily rely on a traditional peer-to-peer (P2P) mechanism for syncing their state as many other blockchains do."
What do you think?

> **Here's how Arbitrum data flow works:**

Reviewer: ```suggestion
Here's how Arbitrum data flow works:
```
I'd remove the bold formatting from here.

> 1. Batching and Submission:

Reviewer: ```suggestion
1. Batching and submission:
```

>    1. The Sequencer queues transactions and batches them together.

Reviewer: ```suggestion
   1. The sequencer queues transactions and batches them together.
```
Sequencer should be written in lowercase since it's a common name (you can also check the [pattern guide](https://www.notion.so/arbitrum/Pattern-guide-Docs-dafaf8d1ce34433d88c1ef3b4b99c19c?pvs=4#f2bc126c8f2e45bb96e452f7cd2acf2d) for these). Can you change the other ocurrences of this in the rest of the document?

>    2. These batches get submitted to the parent chain:

Reviewer: Whenever you differentiate Rollup and AnyTrust modes in other parts of the document, you place "Rollup" first. I'd do the same here for consistency: Rollup first, and then AnyTrust.

>       1. In Anytrust mode, the Sequencer sends the batch to the Data Availability Server (DAS) and then submits the Data Availability Certificate (DACert) which is returned and generated by the DAS to the parent chain.

Reviewer: I'd either say "the Data Availability Servers (DAS)", or "the Data Availability Committee".
I personally would use the second one.

      2. In Rollup mode, the Sequencer submits the batch of transactions directly to the sequencer inbox contract on the parent chain. (Blob or calldata directly)
2. Node Synchronization:
   1. Upon joining the network, a full node:
      1. In Rollup mode, data is read directly from the parent chain calldata or blobs (depending on how the Sequencer posts the data).

>       2. In Anytrust mode, it reads data from the DACert to verify data availability.

Reviewer: ```suggestion
      2. In Anytrust mode, it checks the DACert to verify data availability and queries the data from the DAC.
```

   2. The node continues to follow this process to catch up with the latest chain height.

>    3. Once caught up, the node receives updates on new Sequencer-queued messages directly from the Sequencer feed.(We will provide a detailed view of this at the next section)

Reviewer: ```suggestion
   3. Once caught up, the node receives updates on new Sequencer-queued messages directly from the Sequencer feed (we will provide a detailed view of this process in the next section).
```

3. Catching Up:
   1. If a node falls behind the chain, it reverts to the process described in step 2 to resynchronize with the latest state.

In essence, Arbitrum nodes prioritize data retrieval from the parent chain and rely on the Sequencer for real-time updates, deviating from the traditional P2P synchronization approach used by Ethereum nodes.

> ## **How full nodes decode the data from the parent chain:**

Reviewer: ```suggestion
## How full nodes decode the data from the parent chain:
```
No need to format this as bold, otherwise it will only format the right sidebar.

> Arbitrum full nodes decode data received from the parent chain (Ethereum) to update their local state. This process involves monitoring events, parsing data, and Message Processing.

Reviewer: ```suggestion
Arbitrum full nodes decode data received from the parent chain (and, in the case of AnyTrust chains, the DAC) to update their local state. This process involves monitoring events, parsing data, and processing messages.
```

> 1. Event Querying:

Reviewer: ```suggestion
1. Event querying:
```

   1. Full nodes subscribe to the `SequencerBatchDelivered` event emitted by the inbox contract on the parent chain. This event signifies the arrival of a new batch of transactions.

> 2. Event Parsing:

Reviewer: ```suggestion
2. Event parsing:
```

   1. Upon receiving the `SequencerBatchDelivered` event, the node parses the event data into a `SequencerInboxBatch` struct. This struct typically includes:
      1. `BlockHash`: The hash of the parent chain block containing the batch.
      2. `ParentChainBlockNumber`: The block number of the parent chain block.
      3. `SequenceNumber`: The sequence number of the batch.
      4. `TimeBounds`: Time constraints for the batch.
      5. `AfterDelayedAcc`: Accumulator hash after processing delayed messages.
      6. `AfterDelayedCount`: Count of delayed messages.
      7. `rawLog`: The raw event log data.

> 3. Data Serialization:

Reviewer: ```suggestion
3. Data serialization:
```

   1. The `SequencerInboxBatch` struct serializes into a byte array.
   2. The serialized data adheres to a specific format:
      1. `TimeBounds.MinTimestamp` (8 bytes)
      2. `TimeBounds.MaxTimestamp` (8 bytes)
      3. `TimeBounds.MinBlockNumber` (8 bytes)
      4. `TimeBounds.MaxBlockNumber` (8 bytes)
      5. `AfterDelayedCount` (8 bytes)
      6. `payload` (variable length)
         1. The `payload` field further contains the following:

>             1. **Type:** Indicates the type of payload (e.g., DAS, blob message).

Reviewer: ```suggestion
            1. **Type:** Indicates the type of payload (e.g., DACert, blob message).
```

>             2. **Content:** The actual data associated with the payload type (e.g., DACert, BlobHashes, brotli compressed data).

Reviewer: ```suggestion
            2. **Content:** The actual data associated with the payload type (e.g., DACert hash, BlobHashes, brotli compressed data).
```

Author: We keep all DACert on chain, not only DACert hash, so I can we are good to use DACert here.

Reviewer: Yeah, sorry. I added "hash" to differentiate it from the "Type" in point 1 (it also says DACert). But I think the context is clear, and adding "hash" is worse. So disregard 🙏

> 4. Data Decoding and Retrieval:

Reviewer: ```suggestion
4. Data decoding and retrieval:
```

   1. Based on the `payload` type:

>       1. **DAS Type:** The node queries the Data Availability Service (DAS) to retrieve the raw data.

Reviewer: ```suggestion
      1. **DACert type:** The node queries a Data Availability Service (DAS) to retrieve the raw data.
```
Or, the alternative "queries the Data Availability Committee (DAC)..."

Author: We call this in nitro source code is DASMessageHeader, so maybe I change this to   DAS Message Type for better?

Author: I change all message type to message header instead as it should be more clear

Reviewer: Awesome, yes 💪

>       2. **Blob Message Type:** The node decodes the blob message to obtain the raw data.

Reviewer: ```suggestion
      2. **Blob message type:** The node decodes the blob message to obtain the raw data.
```
Can you also remove the uppercase of the words in the next points? 🙏

      3. **Brotli Message Type:** No extra steps are needed here; continue to the next step.

>    2. **Data Decompression:** If the raw data is Brotli-compressed, the node decompresses it. It's worth noting that the raw data we get from above i and ii might also be Brotli-compressed data.

Reviewer: ```suggestion
   2. Data decompression: If the raw data is Brotli-compressed, the node decompresses it. It's worth noting that the raw data we get from above i and ii might also be Brotli-compressed data.
```

> 5. Message Processing:

Reviewer: ```suggestion
5. Message processing:
```

   1. After decoding and decompressing the data, the node obtains a series of messages.

      1. Message Types:

>          | BatchSegmentKindL2Message            | This message will contain raw data on a series of transactions. Usually, this is a single block. |

Reviewer: You should include an explicit header for the table. Otherwise this first line will be considered the header and it will be formatted differently.

         | ------------------------------------ | ------------------------------------------------------------------------------------------------ |
         | BatchSegmentKindL2MessageBrotli      | The message is the same as the above one, but this is brotli compressed data.                    |
         | BatchSegmentKindDelayedMessages      | This message contains a new delayed message read from the parent chain delayed inbox.            |

>          | BatchSegmentKindAdvanceTimestamp     | This message will notify STF to advance a second of the timestamp state.                         |

Reviewer: ```suggestion
         | BatchSegmentKindAdvanceTimestamp     | This message will notify the State Transition Function (STF) to advance a second of the timestamp state.                         |
```
Since it's the first time that the acronym STF is used. Is that accurate, btw?

         | BatchSegmentKindAdvanceL1BlockNumber | This message will notify STF to advance a new parent chain block number.                         |

>    2. **State Transition:** Finally, the State Transition Function (STF) processes these messages, and the STF will follow the rules to execute and update the Arbitrum node's local state.

Reviewer: ```suggestion
   2. State transition: finally, the State Transition Function (STF) processes these messages, and the STF will follow the rules to execute and update the Arbitrum node's local state.
```

> **Conclusion:**

Reviewer: I think we can remove this block since it doesn't add more value.

This process ensures that Arbitrum nodes can trustless sync an accurate view of the chain without trusting other full nodes on the network.

> ## **How full nodes sync the data from the sequencer feed:**

Reviewer: ```suggestion
## How full nodes sync the data from the sequencer feed
```

Once Arbitrum full nodes have caught up with the chain, they switch from initial synchronization to a real-time update mode. This switch involves receiving data from the sequencer feed, which continuously broadcasts updates about newly queued transactions.

> 1. **Data Acquisition:**

Reviewer: ```suggestion
1. Data acquisition:
```
We didn't format the title of bullet points in bold in the previous section, so to keep consistency, we can remove the bold format here.
Also, remove uppercases from words that are not the first one.
Can you do the same with the next points? 🙏
